File: mobile.fs
#light

//Declare mobiles structure
type Mobile =
    // Declare a mobile that has a ball
    | WithBall of int * int
    // Declare a mobile that has a strut
    | WithStruts of int * ((int * Mobile) * (int * Mobile));;

//Cord and ball mobile
let mobile1 = WithBall(5, 5);;

//Class example
let mobile2 = WithStruts(2, (
  (20, WithStruts(1, (
    (5, WithBall(1, 5)),
    (5, WithStruts(1, (
      (4, WithBall(5, 3)),
      (6, WithBall(1, 2))
    )))
   ))), 
  (10, WithBall(5, 20))
));;

//Unbalanced example
let mobile3 = WithStruts(2, (
  (20, WithStruts(1, (
    (5, WithBall(1, 5)),
    (5, WithStruts(1, (
      (4, WithBall(5, 3)),
      (6, WithBall(1, 2))
    )))
  ))), 
  (10, WithBall(5, 21))
));;

//Finds total weight of passed mobile
let rec totalWeight(mobile : Mobile) =
  //Match mobile structure to sub-type
  match mobile with
  //When cord and ball, return weight
  | WithBall(_, weight) -> weight
  //When cord and strut, return weight of left and right added
  | WithStruts(_, ((_, left), (_, right))) -> totalWeight(left) + totalWeight(right);;
   
//Finds total height of passed mobile
let rec totalHeight(mobile : Mobile) =
  //Match mobile structure to sub-type
  match mobile with
  //When cord and ball, return height
  | WithBall(height, _) -> height
  //When cord and strut, return height 
  | WithStruts(height, ((_, left), (_, right))) -> height + 
    //plus the greater of left and right
    if totalHeight(left) > totalHeight(right) 
    then totalHeight(left) 
    else totalHeight(right);;

//Figures out if passed mobile is balanced
let rec isBalanced(mobile : Mobile) =
  //Match mobile structure to sub-type
  match mobile with
  //When cord and ball, mobile is balanced
  | WithBall(_, _) -> true
  //When cord and strut return true if
  | WithStruts(_, ((leftWidth, left), (rightWidth, right))) -> 
    //left and right are balanced
    isBalanced(left) && isBalanced(right)
      //and the left width * left weight is = to the right width * right weight 
      && leftWidth * totalWeight(left) = rightWidth * totalWeight(right)

printfn "displaying results for cord and ball";;
printfn "total weight is %d" (totalWeight(mobile1));;
printfn "total height is %d" (totalHeight(mobile1));;
printfn "is balanced? %b" (isBalanced(mobile1));;
printfn "displaying results for class example";;
printfn "total weight is %d" (totalWeight(mobile2));;
printfn "total height is %d" (totalHeight(mobile2));;
printfn "is balanced? %b" (isBalanced(mobile2));;
printfn "displaying results for unbalanced example";;
printfn "total weight is %d" (totalWeight(mobile3));;
printfn "total height is %d" (totalHeight(mobile3));;
printfn "is balanced? %b" (isBalanced(mobile3));;

open System
//Pause input
Console.ReadKey(true)

File: mobile.pl
#!/usr/bin/perl -w
#
# Version:
# $Id$
# 
# author name: Allen Madsen
# description:
#    Displays the weight, height, and weather or not a mobile is ballanced
#    for each of the mobiles defined in @mobiles.
#    Mobile data structure is defined by the following psuedo free grammar:
#    C -> [N, S|B]      #Cord
#    S -> P,P           #Strut
#    P -> [N, C]        #Strut Part - Right or Left
#    B -> N             #Ball
#    N -> /[1-9]\d*/    #Number
# ###################################################################

@mobiles = (
  #sample given
  [2,[20,[1,[5,[1,5]],[5,[1,[4,[5,3]],[6,[1,2]]]]]],[10,[5,20]]],
  #chord and ball
  [9, 20],
  #mobile with 1 strut with same submobiles on each side
  [5, [3, [2, 10]], [3, [2, 10]]],
  #unballanced mobile
  [1,[10,[7,7]],[10,[8,8]]],
  #sample given modified to be unballanced
  [2,[20,[1,[5,[8,5]],[5,[1,[3,[5,3]],[6,[1,2]]]]]],[3,[3,20]]],
  #ballanced top, unballanced submobile
  [5, [3, [2, [5, [2, 10]], [3, [2, 10]]]], [3, [2, 20]]],
);

#starts execution
&main(@mobiles);
sub main{
  my $i = 0;
  #for all the mobiles
  foreach my $mobile (@_){
    ++$i;
    print "Mobile $i\n";
    #calculate the info
    my $height = &total_height($mobile);
    my $weight = &total_weight($mobile);
    my $ballanced = &is_ballanced($mobile);
    #and show it
    print "height: $height\nweight: $weight\nballanced: ";
    if($ballanced){
      print "yes";
    } else {
      print "no";
    }
    print "\n\n";
  }
}

sub total_height{
  my $mobile = $_[0];
  #when we have a chord and a ball
  if(@$mobile == 2){
    #return the chord height
    return $mobile->[0];
  } else { #when we have a chord and strut
    #find the left height
    my $left_height = &total_height($mobile->[1][1]);
    #and right height
    my $right_height = &total_height($mobile->[2][1]);
    #decide which is bigger and return that one added to 
    #the current height
    if($left_height > $right_height){
      return $left_height + $mobile->[0];
    } else {
      return $right_height + $mobile->[0];
    }
  }
}

sub total_weight{
  my $mobile = $_[0];
  #when we have a chord and a ball
  if(@$mobile == 2){
    #return the ball weight
    return int($mobile->[1]);
  } else { #when we have a chord and a strut
    #get the total weight of the left and right and return them
    return &total_weight($mobile->[1][1]) + &total_weight($mobile->[2][1]); 
  }
}

sub is_ballanced{
  my $mobile = $_[0];
  #when we have a chord and a ball
  if(@$mobile == 2){
    #it is always ballanced so return 1
    return 1;
  } else { #when we have a chord and a strut
    #get the left weight
    my $left_weight = &total_weight($mobile->[1][1]);
    #and the right weight
    my $right_weight = &total_weight($mobile->[2][1]);
    #and the make sure this level is ballanced and the 
    #left and right are ballanced as well;
    return $left_weight * $mobile->[1][0] == $right_weight * $mobile->[2][0] 
      && &is_ballanced($mobile->[1][1]) && &is_ballanced($mobile->[2][1]); 
  }
}

# ###################################################################
# Revision History:
# $Log$
# 
__END__

=head1 NAME
mobiles.pl -- displays height, weight, and weather or not the
              mobile is ballanced

=head1 SYNOPSIS
mobiles.pl

=head1 DESCRIPTION
Program displays height, weight, and weather or not the
mobile is ballanced

=head1 AUTHOR
Allen Madsen (acm1546@rit.edu)

=head1 README
mobiles.pl filename

=head1 PREREQUISITES
none.